test/mocks/http/mocks.cc

```diff
@@ -160,5 +165,94 @@ MockInstance::MockInstance() {}
+  // Get all headers with matching keys.
+  std::vector<absl::string_view> values;
+  std::pair<std::string, std::vector<absl::string_view>*> context =
+      std::make_pair(key_.get(), &values);
+  Envoy::Http::HeaderMap::ConstIterateCb get_headers_cb = [](const Envoy::Http::HeaderEntry& header,
+                                                             void* context) {
+    auto* typed_context =
+        static_cast<std::pair<std::string, std::vector<absl::string_view>*>*>(context);
+    if (header.key().getStringView() == typed_context->first) {
+      typed_context->second->push_back(header.value().getStringView());
+    }
+    return Envoy::Http::HeaderMap::Iterate::Continue;
+  };
+  headers.iterate(get_headers_cb, &context);
+
+  if (values.empty()) {
+    *listener << "which has no '" << key_.get() << "' header";
+    return false;
+  } else if (values.size() > 1) {
+    *listener << "which has " << absl::StrCat(values.size()) << " '" << key_.get()
+              << "' headers, with values: " << absl::StrJoin(values, ", ");
+    return false;
+  }
+  absl::string_view value = values[0];
+  *listener << "which has a '" << key_.get() << "' header with value " << value << " ";
+  return testing::ExplainMatchResult(matcher_, value, listener);
+}
+
+void HeaderValueOfMatcher::DescribeTo(std::ostream* os) const {
+  *os << "has a '" << key_.get() << "' header with value that "
+      << testing::DescribeMatcher<absl::string_view>(matcher_);
+}
+
+void HeaderValueOfMatcher::DescribeNegationTo(std::ostream* os) const {
+  *os << "doesn't have a '" << key_.get() << "' header with value that "
+      << testing::DescribeMatcher<absl::string_view>(matcher_);
+}
+
+namespace {
+
+class IsSubsetOfHeadersMatcher : public MatcherInterface<const HeaderMap&> {
+public:
+  explicit IsSubsetOfHeadersMatcher(const HeaderMap& expected_headers)
+      : expected_headers_(expected_headers) {}
+
+  bool MatchAndExplain(const HeaderMap& headers, MatchResultListener* listener) const override {
+    // Collect header maps into vectors, to use for IsSubsetOf.
+    auto get_headers_cb = [](const HeaderEntry& header, void* headers) {
+      static_cast<std::vector<std::pair<absl::string_view, absl::string_view>>*>(headers)
+          ->push_back(std::make_pair(header.key().getStringView(), header.value().getStringView()));
+      return HeaderMap::Iterate::Continue;
+    };
+    std::vector<std::pair<absl::string_view, absl::string_view>> arg_headers_vec;
+    headers.iterate(get_headers_cb, &arg_headers_vec);
+    std::vector<std::pair<absl::string_view, absl::string_view>> expected_headers_vec;
+    expected_headers_.iterate(get_headers_cb, &expected_headers_vec);
+
+    return ExplainMatchResult(IsSubsetOf(expected_headers_vec), arg_headers_vec, listener);
+  }
+
+  void DescribeTo(std::ostream* os) const override {
+    *os << "is subset of headers:\n" << expected_headers_;
```

Comment:
nit: Is it more readable as "is a subset of headers" ?

Reply:
Done.